scheds/rust/scx_p2dq/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,1006 @@
+		return NULL;
+	}
+
+	return cpuc;
+}
+
+static __always_inline u64 cpu_dsq_id(int dsq_index, struct cpu_ctx *cpuc) {
+	if (dsq_index < 0 || dsq_index > nr_dsqs_per_llc || dsq_index >= MAX_DSQS_PER_LLC) {
+		scx_bpf_error("invalid dsq index: %d", dsq_index);
+		return 0;
+	}
+	return *MEMBER_VPTR(cpuc->dsqs, [dsq_index]);
+}
+
+
+struct {
+	__uint(type, BPF_MAP_TYPE_ARRAY);
+	__type(key, u32);
+	__type(value, struct llc_ctx);
+	__uint(max_entries, MAX_LLCS);
+} llc_ctxs SEC(".maps");
+
+static struct llc_ctx *lookup_llc_ctx(u32 llc_id)
+{
+	struct llc_ctx *llcx;
+
+	llcx = bpf_map_lookup_elem(&llc_ctxs, &llc_id);
+	if (!llcx) {
+		scx_bpf_error("no llc_ctx for llc %u", llc_id);
+		return NULL;
+	}
+
+	return llcx;
+}
+
+
+struct {
+	__uint(type, BPF_MAP_TYPE_ARRAY);
+	__type(key, u32);
+	__type(value, struct node_ctx);
+	__uint(max_entries, MAX_NUMA_NODES);
+	__uint(map_flags, 0);
+} node_ctxs SEC(".maps");
+
+static struct node_ctx *lookup_node_ctx(u32 node_id)
+{
+	struct node_ctx *nodec;
+
+	nodec = bpf_map_lookup_elem(&node_ctxs, &node_id);
+	if (!nodec) {
+		scx_bpf_error("no node_ctx for node %u", node_id);
+		return NULL;
+	}
+
+	return nodec;
+}
+
+
+struct {
+	__uint(type, BPF_MAP_TYPE_TASK_STORAGE);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+	__type(key, int);
+	__type(value, struct task_ctx);
+} task_ctxs SEC(".maps");
+
+static struct task_ctx *lookup_task_ctx_may_fail(struct task_struct *p)
+{
+	return bpf_task_storage_get(&task_ctxs, p, 0, 0);
+}
+
+static struct task_ctx *lookup_task_ctx(struct task_struct *p)
+{
+	struct task_ctx *taskc = lookup_task_ctx_may_fail(p);
+
+	if (!taskc)
+		scx_bpf_error("task_ctx lookup failed");
+
+	return taskc;
+}
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__type(key, u32);
+	__type(value, u64);
+	__uint(max_entries, P2DQ_NR_STATS);
+} stats SEC(".maps");
+
+static inline void stat_add(enum stat_idx idx, u64 amount)
+{
+	u32 idx_v = idx;
+
+	u64 *cnt_p = bpf_map_lookup_elem(&stats, &idx_v);
+	if (cnt_p)
+		(*cnt_p) += amount;
+}
+
+static inline void stat_inc(enum stat_idx idx)
+{
+	stat_add(idx, 1);
+}
+
+/*
+ * Returns if the task is interactive based on the last DSQ.
+ */
+static bool is_interactive(struct task_ctx *taskc)
+{
+	if (nr_dsqs_per_llc <= 1)
+		return false;
+	return nr_dsqs_per_llc / 2 > taskc->last_dsq_index;
+}
+
+/*
+ * Returns a random llc_ctx
+ */
+static struct llc_ctx *rand_llc_ctx(void)
+{
+	return lookup_llc_ctx(bpf_get_prandom_u32() % nr_llcs);
+}
+
+static bool keep_running(struct cpu_ctx *cpuc, struct task_struct *p)
+{
+	struct llc_ctx *llcx;
+	int i;
+
+	// Only tasks in the least non interactive DSQ can keep running
+	if (cpuc->dsq_index != nr_dsqs_per_llc - 1 ||
+	    p->scx.flags & SCX_TASK_QUEUED ||
+	    cpuc->ran_for >= max_exec_ns)
+		return false;
+
+	if (!(llcx = lookup_llc_ctx(cpuc->llc_id)))
+		return false;
+
+	int nr_queued = 0;
+	bpf_for(i, 0, nr_dsqs_per_llc) {
+		// ignore interactive tasks
+		if (i == 0)
+			continue;
```

Comment:
This doesn't match the `is_interactive` check. That says if there is 1 queue per LLC nothing is interactive, whereas this says if there is 1 queue per LLC everything is interactive. Maybe that's okay but it's not clear why.

Reply:
This is only for the case to allow a task to keep running beyond it's slice. I guess I was thinking that before a task should be allowed to run for more than it's slice it should at least have completed the most interactive slice.